eth/db/header.py

```diff
@@ -211,6 +256,9 @@ def _persist_header_chain(
             rlp.encode(curr_chain_head),
         )
         score = cls._set_hash_scores_to_db(db, curr_chain_head, score)
+        gap_change, gaps = cls._update_header_chain_gaps(db, curr_chain_head)
+        if gap_change is GapChange.GapShrink or gap_change is GapChange.GapSplit:
+            cls._add_block_number_to_hash_lookup(db, curr_chain_head)
```

Comment:
>This assertion is made when persisting blocks that ore older than already persisted canonical blocks.

getting nitpicky, just "being older" doesn't qualify. One can write a bunch of older headers for a parallel history and it won't trigger this write. There has to be a known gap to be written to in order to trigger it.

I has led me to the following elaboration:
> The caller implicitly asserts that persisted headers are canonical here.
This assertion is made when persisting headers that are known to be part of a gap
in the canonical chain. While we do validate that the headers are valid, we can not
be 100 % sure that they will eventually lead to the expected child that fills the gap.
E.g. there is nothing preventing one from persisting an uncle as the final header to
close the gap, even if that will not be the parent of the next consecutive header.
Py-EVM makes the assumption that client code will check and prevent such a scenario.

That said, now that I wrote all this and now that we have `GapChange.GapFilled`, wouldn't it be better to simply make that extra check in Py-EVM when we fill a gap. I mean, load the existing known child (the checkpoint header) and if the ends don't match up, raise an exception. It would then be Trinitys job to handle this which probably means to just log a warning and try to close the gap another time (under the assumption that someone tried to feed us an uncle for whatever reason)

Reply:
> now that we have `GapChange.GapFilled`, wouldn't it be better to simply make that extra check in Py-EVM when we fill a gap. I mean, load the existing known child (the checkpoint header) and if the ends don't match up, raise an exception.

Yup, that seems reasonable to me. So then we have to handle this scenario, with a series of inserted headers between A0 and A4, but the B branch is actually uncled. The top row is the presumed canonical one.

```
A0 -> __ -> __ -> __ -> A4

A0 -> B1 -> __ -> __ -> A4

A0 -> B1 -> B2 -> __ -> A4

A0 -> B1 -> B2 -> __ -> A4  # Tries to insert B3, but it's not a parent of A4, and raises an exception

# client somehow finds the A chain and starts inserting it
# it is deemed non-canonical at first (current PR functionality)
A0 -> B1 -> B2 -> __ -> A4
   \_ A1

A0 -> B1 -> B2 -> __ -> A4
   \_ A1 -> A2

# When inserting A3, pyevm notices that it fills a gap, and links with the child
# It notices that A3's parent is present but non-canonical, and makes 
# B* non-canonical and A[1-3] canonical
A0 -> A1 -> A2 -> A3 -> A4
   \_ B1 -> B2
```

Reply:
Good point! I added a test for that (which also asserts a raised exception in case we are trying to close the gap with `B3`. I found it simpler to cover that in one test with two assertions)